### ISSUE TITLE
Calling REFPROP with VBA

### ISSUE THREAD

**Issue by bkrzn** [usnistgov/REFPROP-wrappers#456]
I'm having a hard time calling REFPROP via VBA. Refprop works fine while working on spreadsheets.
But im desperately trying to call it within VBA.  I also used (or tried using) the subroutine which was posted in #394 but it still doesn´t work. When im trying to call REFPROP an error occurs which states: "sub or function not defined".
 I'm quite a beginner so there could be a chance that i overlook a minor mistake.

Hopefully you Guys can help me.

**REFPROP Version:** [REFPROP10]  
**Operating System and Version:** [Windows 10]  
**Access Method:** [Excel]  
![image](https://user-images.githubusercontent.com/106114590/169898068-ca38a96d-a922-459f-9fe6-1dea26a7d92b.png)

**Comment by ianhbell** [comment]
@EricLemmon Any idea?

**Comment by EricLemmon** [comment]
You've labeled your routine as a subroutine, but it should be a function.  Something like this:

Function DoThis(_...xxx..._)
  DoThis=Refpropvba...
End Function

Then on your spreadsheet you can simply type in "=DoThis(_...xxx..._)"  where the _xxx_ are the numbers that you want to send to the routine.

I also may not have understood the problem correctly, let me know if so.

**Comment by bkrzn** [comment]
Thanks for the quick feedback. I think I have described my problem too unclearly. The command =REFPROP(.....) on the spreadsheet works fine.
I would like to have refprop calculate the properties like in the thread # within my VBA code and pass it to the spreadsheet first after several loops and procedures. I just can't get it to work though. Even with the subroutine (see below) which you specified in the above thread it does not work.

I have also tried your suggestion but this does not work either.

`Sub REFPROPvba(hfld As String, hIn As String, hOut As String, iUnits As Long, iMass As Long, iFlag As Long, a As Double, b As Double, z() As Double, output() As Double, hUnits As String, q As Double, ierr As Long, herr As String)
  Dim hFld2 As String * 10000, hIn2 As String * 255, hOut2 As String * 255, hUnits2 As String * 255, herr2 As String * 255
  Dim iUnits2 As Long, iMass2 As Long, iFlag2 As Long, iUCode2 As Long, ierr2 As Long
  Dim a2 As Double, b2 As Double, z2(ncmax) As Double, X(ncmax) As Double, Y(ncmax) As Double, x3(ncmax) As Double, q2 As Double, Output2(iPropMax) As Double

  hFld2 = hfld: hIn2 = hIn: hOut2 = hOut: iUnits2 = iUnits: iMass2 = iMass: iFlag2 = iFlag: a2 = a: b2 = b
  For i = 1 To ncmax: z2(i) = z(i): Next
  Call REFPROPdll(hFld2, hIn2, hOut2, iUnits2, iMass2, iFlag2, a2, b2, z2(1), Output2(1), hUnits2, iUCode2, X(1), Y(1), x3(1), q2, ierr2, herr2, 10000&, 255&, 255&, 255&, 255&)
  For i = 1 To iPropMax: output(i) = Output2(i): Next
  hUnits = hUnits2: iUCode = iUCode2: q = q2: ierr = ierr2: herr = herr2
  If Trim(hFld2) <> "" Then For i = 1 To ncmax: z(i) = z2(i): Next
End Sub`

**Comment by bkrzn** [comment]
I solved the problem...... I assumed that I can call the REFPROP routine without copying the code from REFPROP in my Project.
Like I said in the beginning: "I'm quite a beginner so there could be a chance that i overlook a minor mistake".

But still thank you for your fast response and help.

**Comment by ryan-w-elliott** [comment]
bkrzn - I don't understand your comment about copying REFPROP into your project. With the REFPROP add in that comes with the NIST software I was assuming you don't need to copy the entire code into your project. Am I mistaken? 
Function DoThis(...xxx...)
DoThis=Refpropvba...
End Function

Should the .xxx... be = (hFld As String, hIn As String, hOut As String, iUnits As Long, iMass As Long, iFlag As Long, a As Double, b As Double, z() As Double, Output() As Double, hUnits As String, q As Double, ierr As Long, herr As String) in the function?

**Comment by ianhbell** [comment]
You might need to (or want to) copy routines from the REFPROP add-in into your own code.

**Comment by JanSukup** [comment]
@bkrzn can you please describe how you solved your issue? I am experiencing the same problem.

**Comment by bkrzn** [comment]
I copied the whole sourcecode from the module "Refprop10Code" from the EXCEL-Wrapper into my own module before any of my own subroutines. Then i could just use

Var1 = REFPROP("P", CarbonDioxide, "TQ", "SI with C", T, 1, 0) 

to call REFPROP.

**Comment by smfamily11** [comment]
> I copied the whole sourcecode from the module "Refprop10Code" from the EXCEL-Wrapper into my own module before any of my own subroutines. Then i could just use
> 
> Var1 = REFPROP("P", CarbonDioxide, "TQ", "SI with C", T, 1, 0)
> 
> to call REFPROP.

I just followed the way you found out, and it is eventually solved to me, too. And this additional comment hopely to help the others facing to the same problem.

1) Copying the whole lines in the module with the name, "Refpro10Code" in VBA, which is included in the original spreadsheet
    provided by NIST.
2) Adding the subroutine for properties that you want to find, such as:
   Sub Example()
         Range("[Cell number]") = REFPROP("H", "WATER", "PT', "SI", Pressure value, Temperature value)
   End Sub
   --------> The result with the example subroutine will be appeared in the cell at which you set.

I have tried to compile, run and check the code it self and using macro with the command button in the UserForm, and both 
ways were successfully worked.
